src/Entity/PageManager.php

```diff
@@ -71,50 +83,24 @@ public function createWithDefaults(array $defaults = []): PageInterface
         return $page;
     }
 
+    /**
+     * NEXT_MAJOR: keep only $this->fixPageUrl->fixUrl$page) in this method.
+     */
     public function fixUrl(PageInterface $page): void
     {
-        if ($page->isInternal()) {
-            $page->setUrl(null); // internal routes do not have any url ...
-
-            return;
-        }
-
-        // hybrid page cannot be altered
-        if (!$page->isHybrid()) {
-            $parent = $page->getParent();
-
-            if (null !== $parent) {
-                $slug = $page->getSlug();
-
-                if (null === $slug) {
-                    $slug = $this->slugify->slugify($page->getName() ?? '');
-
-                    $page->setSlug($slug);
-                }
-
-                $parentUrl = $parent->getUrl();
-
-                if ('/' === $parentUrl) {
-                    $base = '/';
-                } elseif (!str_ends_with($parentUrl ?? '', '/')) {
-                    $base = $parentUrl.'/';
-                } else {
-                    $base = $parentUrl;
-                }
-
-                $url = $page->getCustomUrl() ?? $slug;
-                $page->setUrl('/'.ltrim($base.$url, '/'));
-            } else {
-                $page->setSlug(null);
-
-                $url = $page->getCustomUrl() ?? '';
-                $page->setUrl('/'.ltrim($url, '/'));
-            }
+        $pageFixer = $this->pageFixer;
+
+        if ($pageFixer instanceof SlugifyInterface) {
+            @trigger_error(\sprintf(
+                'Inject %s in %s is deprecated since version 4.10.0 and will be removed in 5.0, use %s instead of.',
+                SlugifyInterface::class,
+                self::class,
+                PageFixerInterface::class,
+            ), \E_USER_DEPRECATED);
+            $pageFixer = new PageFixerService(new AsciiSlugger());
```

Comment:
but this means the `SlugifyInterface` that is passed to the class is not used at all :thinking: seems like a breaking change to me. What if someone customized the slugging behavior?

Reply:
The strategy should be
```
if ($pageFixer instanceof PageFixerInterface) {
     $pageFixer->fixUrl($page);
} else {
     // old implementation
}
```